File: examples/csharp/09_Probabilistic.cs
using StackExchange.Redis;
using System;

class Program
{
    static void Main()
    {
        var redis = ConnectionMultiplexer.Connect("localhost");
        var db = redis.GetDatabase();

        Console.WriteLine("--- 09 God Tier Probabilistic Data Structures ---");

        // 1. HyperLogLog
        db.Execute("PFADD", "hll_cs", "a", "b", "c");
        var count = db.Execute("PFCOUNT", "hll_cs");
        Console.WriteLine($"[HLL] Count: {count}");

        // 2. Cuckoo Filter
        db.Execute("CF.ADD", "cf_cs", "test");
        var exists = db.Execute("CF.EXISTS", "cf_cs", "test");
        Console.WriteLine($"[Cuckoo] Exists: {exists}");

        // 3. Count-Min Sketch
        db.Execute("CMS.INCRBY", "cms_cs", "eventA", 100);
        var q = db.Execute("CMS.QUERY", "cms_cs", "eventA");
        Console.WriteLine($"[CMS] Query: {q}");

        // 4. Top-K
        db.Execute("TOPK.ADD", "topk_cs", "item1", "item2", "item1");
        var list = db.Execute("TOPK.LIST", "topk_cs");
        Console.WriteLine($"[TopK] items: {list}");

        // 5. t-digest
        db.Execute("TDIGEST.ADD", "td_cs", 99.9);
        var p50 = db.Execute("TDIGEST.QUANTILE", "td_cs", 0.5);
        Console.WriteLine($"[t-digest] Quantile: {p50}");
    }
}

File: examples/go/09_probabilistic.go
package main

import (
	"context"
	"fmt"
	"github.com/redis/go-redis/v9"
)

func main() {
	ctx := context.Background()
	rdb := redis.NewClient(&redis.Options{Addr: "localhost:6379"})

	fmt.Println("--- 09 God Tier Probabilistic Data Structures ---")

	// 1. HyperLogLog
	rdb.Do(ctx, "PFADD", "hll_go", "u1", "u2", "u3")
	count, _ := rdb.Do(ctx, "PFCOUNT", "hll_go").Int()
	fmt.Printf("[HLL] Count: %d\n", count)

	// 2. Cuckoo Filter
	rdb.Do(ctx, "CF.ADD", "cf_go", "apple")
	exists, _ := rdb.Do(ctx, "CF.EXISTS", "cf_go", "apple").Int()
	fmt.Printf("[Cuckoo] Exists: %d\n", exists)

	// 3. Count-Min Sketch
	rdb.Do(ctx, "CMS.INCRBY", "cms_go", "view", 10)
	val, _ := rdb.Do(ctx, "CMS.QUERY", "cms_go", "view").Int()
	fmt.Printf("[CMS] Count: %d\n", val)

	// 4. Top-K
	rdb.Do(ctx, "TOPK.ADD", "topk_go", "A", "B", "A", "C", "A")
	list, _ := rdb.Do(ctx, "TOPK.LIST", "topk_go").StringSlice()
	fmt.Printf("[TopK] List: %v\n", list)

	// 5. t-digest
	rdb.Do(ctx, "TDIGEST.ADD", "td_go", 50.5)
	rdb.Do(ctx, "TDIGEST.ADD", "td_go", 100.0)
	q, _ := rdb.Do(ctx, "TDIGEST.QUANTILE", "td_go", 0.5).Float64()
	fmt.Printf("[t-digest] Median: %.2f\n", q)
}
